lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1271,33 +1295,51 @@ def _on_logging_relation_departed(self, event: RelationDepartedEvent):
             event: a `CharmEvent` in response to which the Loki
                 charm must update its relation data.
         """
-        self._update_alert_rules(event.relation)
-        self._check_alert_rules()
+        self.on.loki_push_api_alert_rules_changed.emit(
+            relation=event.relation,
+            relation_id=event.relation.id,
+            app=self._charm.app,
+            unit=self._charm.unit,
+        )
 
-    def _process_logging_relation_changed(self, relation: Relation):
+    def _should_update_alert_rules(self, relation) -> bool:
+        """Determine whether alert rules should be regenerated.
+
+        If there are alert rules in the relation data bag, tell the charm
+        whether or not to regenerate them based on the boolean returned here.
+        """
+        if relation.data.get(relation.app).get("alert_rules"):
+            return True
+        return False
+
+    def _process_logging_relation_changed(self, relation: Relation) -> bool:
         """Handle changes in related consumers.
 
         Anytime there are changes in relations between Loki
         and its consumers charms, Loki set the `loki_push_api`
-        into the relation data.
+        into the relation data. Set the endpoint building
+        appropriately, and if there are alert rules present in
+        the relation, let the caller know.
         Besides Loki generates alert rules files based what
         consumer charms forwards,
 
         Args:
             relation: the `Relation` instance to update.
+
+        Returns:
+            A boolean indicating whether an event should be emitted so we
+            only emit one on lifecycle events
```

Comment:
"event should be emitted ..." is a separate (external) concern. Move the comment there?